Move `Option<Box<_>>` into `ExtraInstructionAttributes`.

kevinhartman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
The primary motivation here is to encapsulate the hairy logic of managing memory for extra attributes. By moving the `Option<Box<_>>` into the `ExtraInstructionAttributes` struct itself, clients are no longer burdened with the responsibility of checking if the `Option` contains a value, unwrapping it etc.; the semantics of that aren't relevant to clients like `DAGCircuit` and `CircuitData`. Instead, the `ExtraInstructionAttributes` acts as a container for optional attributes, and can easily drop the `Box` internally if all attributes are cleared.

A give-away that this pattern could be helpful was `ExtraInstructionAttributes::new`'s return type, which was previously `Option<Self>`.

### Details and comments
The size of `ExtraInstructionAttributes` (now held by value by instructions) should be the same size as the old `Option<Box<ExtraInstructionAttributes>>`, so this should not have memory implications regarding the size of instructions.

Note that `ExtraAttrs` is completely private to the `circuit_instruction` module, i.e. it's only an implementation detail and in no way part of the interface.

kevinhartman committed:
Move Option<Box<_>> into ExtraInstructionAttributes.

Previously, CircuitInstruction and PackedInstruction held
the ExtraInstructionAttributes struct within an Option<Box<_>>.
By putting the Option<Box<_>> inside ExtraInstructionAttributes,
we can use the struct itself to manage its memory and provide
access to the attributes behind a nicer interface.

The size of ExtraInstructionAttributes should be the same size
as the old Option<Box<ExtraInstructionAttributes>>, so this
should not have memory implications.

jakelishman commented on crates/circuit/src/circuit_instruction.rs:
```diff
@@ -30,16 +30,21 @@ use crate::operations::{
+    duration: Option<PyObject>,
+    unit: Option<String>,
+    condition: Option<PyObject>,
+}
```
Two minor nits:

- imo for naming it'd be good to be consistent with either "attrs" or "attributes" between the two structs.  I know it's just internal to the module (hence just being a nit) and that we used `extra_attrs` as the attribute name, haha
- can you add a replacement doc comment explaining why the split exists, for the future?

kevinhartman replied:
Done in bca33aa9443b0122a8411e057e40f63b7d5eb59f.

jakelishman commented on crates/circuit/src/circuit_instruction.rs:
```diff
@@ -30,16 +30,21 @@ use crate::operations::{
-    pub unit: Option<String>,
-    pub condition: Option<PyObject>,
+    attributes: Option<Box<ExtraAttrs>>,
 }
```
Any particular reason to use a record struct rather than a tuple struct?

kevinhartman replied:
Nope! I thought of that later, but I think it makes sense.

Done in bca33aa9443b0122a8411e057e40f63b7d5eb59f.

jakelishman commented on crates/circuit/src/circuit_instruction.rs:
```diff
@@ -51,32 +56,111 @@ impl ExtraInstructionAttributes {
+        self.attributes
+            .as_deref()
+            .and_then(|attrs| attrs.unit.as_deref())
+    }
```
Do you want to put the setters for `duration` and `unit` on now, or wait til we need them?

kevinhartman replied:
Done in bca33aa9443b0122a8411e057e40f63b7d5eb59f.

jakelishman commented on crates/circuit/src/operations.rs:
```diff
@@ -425,29 +425,25 @@ impl StandardGate {
+            out = out.call_method0(py, "to_mutable")?;
+            out.setattr(py, "condition", condition)?;
         }
+        Ok(out)
```
This is a potential perf regression, because it now always builds a Python `dict` when before it would attempt to skip it.  Not huge, but probably better to keep avoiding it.

kevinhartman replied:
Good point.

Resolved in 943b14260bb08c4413ef3a1e3cdae754bb72ab58.

jakelishman reviewed (commented):
Thanks for doing this, it looks like a very sensible idea.

kevinhartman committed:
Address review comments.

- Use tuple struct.
- Use 'Attributes' over 'Attrs'.
- Add doc comment for internal 'ExtraAttributes' struct.
- Add doc comments for methods.
- Add setters for unit and duration.

kevinhartman committed:
Fix performance regression from unnecessary dict creation.

jakelishman reviewed (approved):
Thanks for this and the changes.  I'll leave unmerged because I saw elsewhere that Ray was going to look too.

raynelfss reviewed (approved):
This is all sensible to me, my only comment was about the usage of a map-like struct instead of a simple tuple struct but it seems it's already been corrected :)

This should be good as is. Thanks for working on it! 🚀
